frontend/catalyst/jax_tracer.py

```diff
@@ -529,4 +530,7 @@ def custom_lower_jaxpr_to_module(
                 continue
             op.attributes["llvm.linkage"] = ir.Attribute.parse("#llvm.linkage<internal>")
 
-    return ctx.module, ctx.context
+    # Inject setup and finalize functions.
+    append_modules(ctx.module, ctx.context)
```

Comment:
I see your way of thinking. And separating the JAXPR->MLIR (pre-runtime) pass from MLIR (pre-runtime) -> MLIR (post-runtime) pass is a possibility. I just wanted to contain all MLIR generated / transformations in a tightly coupled space. I have no real preference, but a suitable compromise might be to keep the `append_modules` in this folder but in the caller function.

Reply:
By folder do you mean module? 😅
In that case I'm not so sure about the suggestion, because the code in here should only be concerned with JAX things (hence the name `jax_tracer`, and the init/teardown functions are in no way related to JAX) 🤔
Where in the compiler-related modules I don't care too much, but here it is misplaced I think. (Imo we should be looking to increase the separation between modules with APIs, not reduce them. The core PennyLane repo apparently suffers extensively from this.)

If it means inserted an "MLIR preparation" stage in the compilation pipeline, I'm all for it! (would relate to #48 I think)

Reply:
https://github.com/PennyLaneAI/catalyst/pull/38/commits/c863745abd2d696adbce4cba442c8107b897e0c8 Technically, there's no reason why we can't have two "Compiler" instances. One from JAXPR->MLIR and the other from MLIR->binary. But the separation designing a separation point is a bit arbitrary. And the interfaces needed for the JAXPR->MLIR pipeline might be different. I have given some thought about a more general way to express a unified phase order but haven't come up with something that is as satisfying as I would like. :sweat_smile: